sonar-batch/src/main/java/org/sonar/batch/rule/RuleFinderCompatibility.java

```diff
@@ -48,18 +63,48 @@ public Rule findByKey(String repositoryKey, String key) {
 
   @Override
   public Rule findByKey(RuleKey key) {
-    DefaultActiveRule ar = (DefaultActiveRule) activeRules.find(key);
-    return ar == null ? null : Rule.create(key.repository(), key.rule()).setName(ar.name());
+    return toRule(activeRules.find(key));
   }
 
   @Override
   public Rule find(RuleQuery query) {
-    throw new UnsupportedOperationException("Unable to find rule by query");
+    Collection<Rule> all = findAll(query);
+    if (all.size() > 1) {
+      throw new IllegalArgumentException("Non unique result for rule query: " + ReflectionToStringBuilder.toString(query, ToStringStyle.SHORT_PREFIX_STYLE));
+    } else if (all.size() == 0) {
+      return null;
+    } else {
+      return all.iterator().next();
+    }
   }
 
   @Override
   public Collection<Rule> findAll(RuleQuery query) {
```

Comment:
:new: ![MAJOR](https://raw.githubusercontent.com/henryju/image-hosting/master/major.png)The Cyclomatic Complexity of this method "findAll" is 13 which is greater than 10 authorized.[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)